src/main/extension/extension_install.cpp

```diff
@@ -347,46 +349,82 @@ static unique_ptr<ExtensionInstallInfo> InstallFromHttpUrl(DBConfig &config, con
+		duckdb_httplib::Client cli(url_base.c_str());
+		if (http_logger) {
+			cli.set_logger(http_logger->GetLogger<duckdb_httplib::Request, duckdb_httplib::Response>());
+		}
+
+		duckdb_httplib::Headers headers = {
+		    {"User-Agent", StringUtil::Format("%s %s", config.UserAgent(), DuckDB::SourceID())}};
+
 		if (install_info && !install_info->etag.empty()) {
 			headers.insert({"If-None-Match", StringUtil::Format("%s", install_info->etag)});
 		}
-	}
-
-	auto res = cli.Get(url_local_part.c_str(), headers);
 
-	if (install_info && res && res->status == 304) {
-		return install_info;
-	}
+		res = cli.Get(url_local_part.c_str(), headers);
+		if (install_info && res && res->status == 304) {
+			return install_info;
+		}
 
-	if (!res || res->status != 200) {
-		// create suggestions
-		string message;
-		auto exact_match = ExtensionHelper::CreateSuggestions(extension_name, message);
-		if (exact_match && !ExtensionHelper::IsRelease(DuckDB::LibraryVersion())) {
-			message += "\nAre you using a development build? In this case, extensions might not (yet) be uploaded.";
+		if (res && res->status == 200) {
+			// success!
+			break;
 		}
+		// failure - check if we should retry
+		bool should_retry = false;
 		if (res.error() == duckdb_httplib::Error::Success) {
-			throw HTTPException(res.value(), "Failed to download extension \"%s\" at URL \"%s%s\" (HTTP %n)\n%s",
-			                    extension_name, url_base, url_local_part, res->status, message);
+			switch (res->status) {
+			case 408: // Request Timeout
+			case 418: // Server is pretending to be a teapot
+			case 429: // Rate limiter hit
+			case 500: // Server has error
+			case 503: // Server has error
+			case 504: // Server has error
+				should_retry = true;
+				break;
+			default:
+				break;
+			}
 		} else {
-			throw IOException("Failed to download extension \"%s\" at URL \"%s%s\"\n%s (ERROR %s)", extension_name,
-			                  url_base, url_local_part, message, to_string(res.error()));
+			// always retry on duckdb_httplib::Error::Error
+			should_retry = true;
+		}
+		retry_count++;
+		if (!should_retry || retry_count >= MAX_RETRY_COUNT) {
+			// if we should not retry or exceeded the number of retries - bubble up the error
+			string message;
+			auto exact_match = ExtensionHelper::CreateSuggestions(extension_name, message);
+			if (exact_match && !ExtensionHelper::IsRelease(DuckDB::LibraryVersion())) {
+				message += "\nAre you using a development build? In this case, extensions might not (yet) be uploaded.";
+			}
+			if (res.error() == duckdb_httplib::Error::Success) {
+				throw HTTPException(res.value(), "Failed to download extension \"%s\" at URL \"%s%s\" (HTTP %n)\n%s",
+				                    extension_name, url_base, url_local_part, res->status, message);
+			} else {
+				throw IOException("Failed to download extension \"%s\" at URL \"%s%s\"\n%s (ERROR %s)", extension_name,
+				                  url_base, url_local_part, message, to_string(res.error()));
```

Comment:
Should we concatenate error messages to preserve information?
Like `Failed to download extension x at URL z (HTTP 408, HTTP 408, HTTP 429).`.

Here you get only the last one (that is enough, but maybe it's informative to see multiple failures in having users understand that has been tried hard).

Reply:
I would say we leave this functionally close to httpfs extension and start improving it only after we deduplicate the code again.

Reply:
This code comes from extension_install, so there is only one copy of this, but agree can be reviewed later.